### PR TITLE
Price estimates

### PR DESCRIPTION
**Pull request by Brianna-Flo**
## Description

<what this pull request is doing>

- implement price estimate functionality using the Walmart API
- Display total estimated price of recipe
- Created Recipe Cost modal which presents to user list of and price of each missing ingredient
- Display estimated price of each item on the grocery list view
- Create consistent naming (price -> cost) across variables
- Create loading screen modal
- Update user signed in (set the user) after login
- Implemented feedback regarding fetching http requests
- Allow a user to mark groceries they have purchased and clear any groceries purchased from their list

<what will be done in later PRs and not included here>

- batch fetching to prevent overloading the server
- Further styling/implementation of react component UI library

## Milestones
<the milestones or stories/features that this works towards>

- backend implementation of the app to estimate pricing of grocery list

## Resources
<links to tutorials, code snippets, inspirations>
<if you took or translated specific code from the internet, please call it out here!>
https://rapidapi.com/guides/parallel-api-requests
Axios: https://axios-http.com/docs/api_intro 
https://axios-http.com/docs/example
https://geshan.com.np/blog/2023/11/axios-typescript/

## Test Plan
<insert images or gifs of feature>
<otherwise, say how code was tested if not UI facing>
<any edge cases you might have specifically tested>

- Rounding to 2 decimal places only for decimal values
- Present default value for ingredient if search unsuccessful
- Presenting empty homepage if user not logged in
- Logging in and out without page refresh

<div>
    <a href="https://www.loom.com/share/49931fa4c14a4ee98d8ccd4fa662db41">
      <p>Estimating Grocery List Price - Watch Video</p>
    </a>
    <a href="https://www.loom.com/share/49931fa4c14a4ee98d8ccd4fa662db41">
      
    </a>
  </div>

### CODE REVIEW

File: frontend/src/pages/GroceryList.tsx [hunk 1]
```diff
@@ -35,14 +38,35 @@ const GroceryList: React.FC<GPToggleNavBarProps> = ({ navOpen, toggleNav }) => {
       setGroceryListCost,
     });
   }, []);
+
+  const toggleGroceryCheck = async (groceryItem: string) => {
+    await axios.put(`${databaseUrl}/generateList/check`, {ingredientName: groceryItem}, axiosConfig)
```
Comment (paulshaoyuqiao) [inline frontend/src/pages/GroceryList.tsx:43]: nice!

File: frontend/src/utils/utils.ts [hunk 2]
```diff
@@ -66,6 +66,18 @@ const handleAuthInputChange = (
   setFormData((prev) => ({ ...prev, [credential]: value }));
 };
 
+  const parseGroceryListDepartments = (
+    groceryList: GPRecipeIngredientTypes[]
+  ) => {
+    let departments: string[] = [];
+    for (const grocery of groceryList) {
+      if (!departments.includes(grocery.department)) {
+        departments = [...departments, grocery.department];
```
Comment (paulshaoyuqiao) [inline frontend/src/utils/utils.ts]: can we leverage a set instead of an array so we don't have to use the `includes` check, which is expensive?  ES6/TS guarantees elements are in the insertion order.

File: backend/utils/utils.ts [hunk 3]
```diff
@@ -165,10 +166,36 @@ const createGroceryList = ({
...
 };
 
+type GPEstimateListCostTypes = {
+  ingredientsToPurchase: GPRecipeIngredientTypes[]
+}
+
+const estimateListCost = async ({ingredientsToPurchase}: GPEstimateListCostTypes) => {
+  let estimatedCost = 0;
+  for (const ingredient of ingredientsToPurchase) {
+    const ingredientPrice = await getPriceOfIngredient({ingredient})
```
Comment (paulshaoyuqiao) [inline backend/utils/utils.ts]: do we need to account for the quantity of each ingredient?

File: frontend/src/pages/NewListPage.tsx [hunk 4]
```diff
@@ -98,6 +100,7 @@ const NewListPage: React.FC<GPToggleNavBarProps> = ({ navOpen, toggleNav }) => {
       setMessage({ error: true, message: "Failed to generate grocery list" });
     }
     const data = await response.json()
+    navigate("/grocery-list")
```
Comment (paulshaoyuqiao) [inline frontend/src/pages/NewListPage.tsx]: can we refactor out the navigate string here so it can be reused in other react files too?

File: backend/routes/groceryListRoutes.ts [hunk 5]
```diff
@@ -11,61 +14,125 @@ const prisma = new PrismaClient();
...
-    let ingredientOnHandQuantity = ingredientOnHand.quantity
-    let recipeIngredientQuantity = parseInt(recipeIngredient.quantity)
-    if (ingredientOnHandUnit !== recipeIngredientUnit) {
-        if (ingredientOnHandUnit && recipeIngredientUnit) {
-            //ingredient units converted to convert-unit library compatible format
-            recipeIngredientQuantity = convert(recipeIngredientQuantity).from(recipeIngredientUnit).to(ingredientOnHandUnit)
-            recipeIngredientUnit = ingredientOnHandUnit
-        }
-        // TODO: otherwise, one or both units null, assume user has enough??
+const enoughOnHand = ({
```
Comment (paulshaoyuqiao) [inline backend/routes/groceryListRoutes.ts]: let's rename this? not entirely clear what this function does by its own name

File: backend/routes/groceryListRoutes.ts [hunk 6]
```diff
@@ -11,61 +14,125 @@ const prisma = new PrismaClient();
...
+  recipeIngredient,
+}: GPEnoughOnHandTypes) => {
+  let ingredientOnHandUnit =
+    unitConversions[ingredientOnHand.unit.toLowerCase()];
+  let recipeIngredientUnit =
+    unitConversions[recipeIngredient.unit.toLowerCase()];
+  let ingredientOnHandQuantity = ingredientOnHand.quantity;
+  let recipeIngredientQuantity = parseInt(recipeIngredient.quantity);
+  if (ingredientOnHandUnit !== recipeIngredientUnit) {
+    if (ingredientOnHandUnit && recipeIngredientUnit) {
```
Comment (paulshaoyuqiao) [inline backend/routes/groceryListRoutes.ts]: should the `null` check come after the equality check?
